sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsIO.java

```diff
@@ -0,0 +1,488 @@
+          connection.stop();
+          connection.close();
+        }
+      } catch (Exception e) {
+        throw new IOException(e);
+      }
+    }
+
+  }
+
+  /**
+   * A {@link PTransform} to write to a JMS queue. See {@link JmsIO} for
+   * more information on usage and configuration.
+   */
+  public static class Write extends PTransform<PCollection<String>, PDone> {
+
+    protected ConnectionFactory connectionFactory;
+    protected String queue;
+    protected String topic;
+
+    public Write withConnectionFactory(ConnectionFactory connectionFactory) {
+      return new Write(connectionFactory, queue, topic);
+    }
+
+    public Write withQueue(String queue) {
+      return new Write(connectionFactory, queue, topic);
+    }
+
+    public Write withTopic(String topic) {
+      return new Write(connectionFactory, queue, topic);
+    }
+
+    private Write() {}
+
+    private Write(ConnectionFactory connectionFactory, String queue, String topic) {
+      this.connectionFactory = connectionFactory;
+      this.queue = queue;
+      this.topic = topic;
+    }
+
+    @Override
+    public PDone apply(PCollection<String> input) {
+      input.apply(ParDo.of(new JmsWriter(connectionFactory, queue, topic)));
+      return PDone.in(input.getPipeline());
+    }
+
+    @Override
+    public void validate(PCollection<String> input) {
+      if (connectionFactory == null) {
+        throw new IllegalArgumentException("ConnectionFactory is required");
+      }
+      if (queue == null && topic == null) {
+        throw new IllegalArgumentException("Either queue or topic is required");
+      }
+    }
+
+    private static class JmsWriter extends DoFn<String, Void> {
+
+      private ConnectionFactory connectionFactory;
+      private String queue;
+      private String topic;
+
+      private Connection connection;
+      private Session session;
+      private MessageProducer producer;
+
+      public JmsWriter(ConnectionFactory connectionFactory, String queue, String topic) {
+        this.connectionFactory = connectionFactory;
+        this.queue = queue;
+        this.topic = topic;
+      }
+
+      @Override
+      public void startBundle(Context c) throws Exception {
+        if (producer == null) {
+          this.connection = connectionFactory.createConnection();
+          this.connection.start();
+          this.session = this.connection.createSession(false, Session.AUTO_ACKNOWLEDGE);
+          Destination destination;
+          if (queue != null) {
+            destination = session.createQueue(queue);
+          } else {
+            destination = session.createTopic(topic);
+          }
+          this.producer = this.session.createProducer(destination);
+        }
+      }
+
+      @Override
+      public void processElement(ProcessContext ctx) throws Exception {
+        String value = ctx.element();
+
+        TextMessage message = session.createTextMessage(value);
+        producer.send(message);
```

Comment:
currently in Beam we do not call finish bundle if there is an exception in this code.

To prevent connection leaks, maybe you want wrap this code as follows?

```java
try {
    TextMessage message = ...
    producer.send ...
} catch (RuntimeException | Exception t) {
    finishBundle(null);
    throw t;
}
```

or something like that.

@kennknowles he happens to not use the context in `finishBundle`, but is it easy to turn the `ProcessContext` into a `Context`?